Add Threat Intelligence Section (#7905)
opensearch-project/documentation-website: _security-analytics/threat-intelligence/api/source.md

---
layout: default

> title:  Source 

Author: ```suggestion
title:  Source API
```

parent: Threat intelligence APIs
grand_parent: Threat intelligence
nav_order: 50
---

# Source API

The threat intelligence Source API updates and returns information about tasks related to threat intelligence source configurations.

## Create or update threat intelligence source

Creates or updates a threat intelligence source and loads Indicators of Compromise (IoCs) from that source.

### Path and HTTP methods

```json
POST _plugins/_security_analytics/threat_intel/sources
PUT _plugins/_security_analytics/threat_intel/sources/<source_id>
```

### Request Fields

> | Field  | Type  | Description  |

Reviewer: The table is formatted incorrectly: 
<img width="663" alt="image" src="https://github.com/user-attachments/assets/e95d3a1a-d30b-4d0f-80e5-659e40d66cbd">

Reviewer: Looks good!

| `source.ioc_upload.file_name`  | String   | The name of the file containing IOCs, such as `test`. Applicable for the`IOC_UPLOAD` type.  |
| `source.ioc_upload.iocs`   | Array of Objects | List of IOCs in the `STIX2` format. Applicable for `IOC_UPLOAD` type. |
| `source_config.source.s3`   | Object   | The details about the S3 source.  Applicable for `S3_CUSTOM` type.   |
| `source_config.source.s3.bucket_name` | String  | The name of the S3 bucket, such as `threat-intel-s3-test-bucket`. Applicable for `S3_CUSTOM` type                                                                        |
| `source_config.source.s3.object_key`  | String   | The key of the object in the S3 bucket, such `alltypess3object`. Applicable for `S3_CUSTOM` type.   |
| `source_config.source.s3.region`  | String | The region where the S3 bucket is located. Example: `us-west-2`. Applicable for `S3_CUSTOM` type.  |
| `source_config.source.s3.role_arn`    | String   | The ARN of the role used to access the S3 bucket, such as `arn:aws:iam::248279774929:role/threat_intel_s3_test_role`. Applicable for `S3_CUSTOM` type. |

> #### IOC Fields (STIX2)  

Author: Settling on IOC

                        "schedule": {
                            "interval": {
                                "start_time": 1717097122,
                                "period": 1,
                                "unit": "Days"
                            }
                        },
                        "state": "AVAILABLE",
                        "refresh_type": "FULL",
                        "last_refreshed_time": 1719449576533,
                        "last_refreshed_user": null,
                        "enabled": true,
                        "ioc_types": [
                            "ip",
                            "hash"
                        ]
                    }
                }
            }
        ]
    }
}
```

## Delete threat intelligence source API

Deletes a threat intelligence source

### Path and HTTP methods

```
DELETE /_plugins/_security_analytics/threat_intel/sources/<source-id>
```

### Example request

```
DELETE /_plugins/_security_analytics/threat_intel/sources/2c0u7JAB9IJUg27gcjUp
```
{% include copy-curl.html %}

### Example response

```json
{
  "_id": "2c0u7JAB9IJUg27gcjUp"
}
```
---

> ## Refresh Source 

Reviewer: ```suggestion
## Refresh source 
```

> Downloads any IoCs from the threat intelligence source. Only supports for the `S3_CUSTOM` type source.

Reviewer: ```suggestion
Downloads any IoCs from the threat intelligence source. Only supports the `S3_CUSTOM` type source.
```

---
layout: default
title:  Source API
parent: Threat intelligence APIs
grand_parent: Threat intelligence
nav_order: 50
---

# Source API

The threat intelligence Source API updates and returns information about tasks related to threat intelligence source configurations.

## Create or update threat intelligence source

Creates or updates a threat intelligence source and loads Indicators of Compromise (IOCs) from that source.

### Path and HTTP methods

```json
POST _plugins/_security_analytics/threat_intel/sources
PUT _plugins/_security_analytics/threat_intel/sources/<source_id>
```

> ### Request Fields

Author: ```suggestion
### Request fields
```

| `source_config.source.ioc_upload.file_name` | String   | The name of the file uploaded. Example: `test`. |
| `source_config.source.ioc_upload.iocs`      | Array of Objects  | Any additional information about the IoC upload. When the IoC is stored successfully, this appears as an empty array.   |
| `source_config.enabled`   | Boolean    | Indicates whether the threat intelligence source is enabled.  |
| `source_config.enabled_time`    | String or Null    | The date and time when the source was enabled.   |
| `source_config.last_update_time`  | String (DateTime) | The date and time when the threat intelligence source was last updated.  |
| `source_config.schedule`  | String or Null    | The schedule for the threat intelligence source.  |
| `source_config.state`    | String    | The current state of the threat intelligence source.  |
| `source_config.refresh_type`    | String   | The type of refresh applied to the source.  |
| `source_config.last_refreshed_user`   | String or Null    | The user who last refreshed the source. |
| `source_config.last_refreshed_time`         | String (DateTime) | The date and time when the source was last refreshed. |

### Example requests 

> 

Author: The following examples show how to use the Source API.

> 

Author: ```suggestion
The following examples show how to use the Source API.
```

Reviewer 2: @Naarcha-AWS "example requests show you how"

        },
        {
          "id": "2",
          "name": "uldzafothwgik",
          "type": "hashes",
          "value": "example-has00001",
          "severity": "thvvz",
          "created": "2024-06-20T01:06:20.562008Z",
          "modified": "2024-06-20T02:06:20.56201Z",
          "description": "first one here",
          "labels": [
            "ik"
          ],
          "feed_id": "jl",
          "spec_version": "gavvnespe",
          "version": -4356924786557562654
        }
      ]
    }
  }
}
```
{% include copy-curl.html %}

#### S3_CUSTOM type source

```json
POST _plugins/_security_analytics/threat_intel/sources/
{
 "type": "S3_CUSTOM",
 "name": "example-ipv4-from-SAP-account",
 "format": "STIX2",
 "store_type": "OS",
 "enabled": "true",
 "schedule": {
  "interval": {
   "start_time": 1717097122,
   "period": "10",
   "unit": "DAYS"
  }
 },
 "source": {
  "s3": {
   "bucket_name": "threat-intel-s3-test-bucket",
   "object_key": "alltypess3object",
   "region": "us-west-2",
   "role_arn": "arn:aws:iam::248279774929:role/threat_intel_s3_test_role"
  }
 },
 "ioc_types": [
  "domain-name",
  "ipv4-addr"
 ]
}
```
{% include copy-curl.html %}

### Example responses

> 

Author: ```suggestion

The following example responses show what OpenSearch returns after a successful request.
```

```

### Example request

```
GET /_plugins/_security_analytics/threat_intel/sources/<source-id>
```
{% include copy-curl.html %}

### Example response

```json
{
  "_id": "a-jnfjkAF_uQjn8Weo4",
  "_version": 2,
  "source_config": {
    "name": "my_custom_feed_2",
    "format": "STIX2",
    "type": "S3_CUSTOM",
    "ioc_types": [
      "ipv4_addr",
      "hashes"
    ],
    "description": "this is the description",
    "created_by_user": null,
    "created_at": "2024-06-27T00:52:56.373Z",
    "source": {
      "s3": {
        "bucket_name": "threat-intel-s3-test-bucket",
        "object_key": "bd",
        "region": "us-west-2",
        "role_arn": "arn:aws:iam::540654354201:role/threat_intel_s3_test_role"
      }
    },
    "enabled": true,
    "enabled_time": "2024-06-27T00:52:56.373Z",
    "last_update_time": "2024-06-27T00:52:57.824Z",
    "schedule": {
      "interval": {
        "start_time": 1717097122,
        "period": 1,
        "unit": "Days"
      }
    },
    "state": "AVAILABLE",
    "refresh_type": "FULL",
    "last_refreshed_user": null,
    "last_refreshed_time": "2024-06-27T00:52:56.533Z"
  }
}
```
---

## Search threat intelligence source 

> Searches for threat intelligence source matches based on the search query. The request body expects a search query. For query options, see [Query DSL]({{site.url}}{{site.baseurl}}/query-dsl/).

Author: ```suggestion
Searches for threat intelligence source matches based on the search query. The request body expects a search query. For query options, see [Query DSL]({{site.url}}{{site.baseurl}}/query-dsl/).

```

| `source_config.source.ioc_upload.file_name` | String   | The name of the uploaded file. Example: `test`. |
| `source_config.source.ioc_upload.iocs`      | Array of objects  | Any additional information about the IOC upload. When the IOC is stored successfully, this appears as an empty array.   |
| `source_config.enabled`   | Boolean    | Indicates whether the threat intelligence source is enabled.  |
| `source_config.enabled_time`    | String or null    | The date and time when the source was enabled.   |
| `source_config.last_update_time`  | String (DateTime) | The date and time when the threat intelligence source was last updated.  |
| `source_config.schedule`  | String or null    | The schedule for the threat intelligence source.  |
| `source_config.state`    | String    | The current state of the threat intelligence source.  |
| `source_config.refresh_type`    | String   | The type of refresh applied to the source.  |
| `source_config.last_refreshed_user`   | String or null    | The user who last refreshed the source. |
| `source_config.last_refreshed_time`         | String (DateTime) | The date and time when the source was last refreshed. |

### Example requests 

> The following examples show how to use the Source API.

Author: ```suggestion
The following example requests show how to use the Source API.

```

> The following examples show how to use the Source API.

Reviewer 2: ```suggestion
The following example requests show you how to use the Source API.
```

          "value": "example-has00001",
          "severity": "thvvz",
          "created": "2024-06-20T01:06:20.562008Z",
          "modified": "2024-06-20T02:06:20.56201Z",
          "description": "first one here",
          "labels": [
            "ik"
          ],
          "feed_id": "jl",
          "spec_version": "gavvnespe",
          "version": -4356924786557562654
        }
      ]
    }
  }
}
```
{% include copy-curl.html %}

#### S3_CUSTOM type source

```json
POST _plugins/_security_analytics/threat_intel/sources/
{
 "type": "S3_CUSTOM",
 "name": "example-ipv4-from-SAP-account",
 "format": "STIX2",
 "store_type": "OS",
 "enabled": "true",
 "schedule": {
  "interval": {
   "start_time": 1717097122,
   "period": "10",
   "unit": "DAYS"
  }
 },
 "source": {
  "s3": {
   "bucket_name": "threat-intel-s3-test-bucket",
   "object_key": "alltypess3object",
   "region": "us-west-2",
   "role_arn": "arn:aws:iam::248279774929:role/threat_intel_s3_test_role"
  }
 },
 "ioc_types": [
  "domain-name",
  "ipv4-addr"
 ]
}
```
{% include copy-curl.html %}

### Example responses

The following example responses show what OpenSearch returns after a successful request.

> The following example responses show what OpenSearch returns after a successful request.

Author: ```suggestion

```
